File: application/src/com/ferg/awfulapp/task/QuoteRequest.java
package com.ferg.awfulapp.task;

import android.content.ContentUris;
import android.content.ContentValues;
import android.content.Context;
import android.net.Uri;
import com.ferg.awfulapp.constants.Constants;
import com.ferg.awfulapp.provider.AwfulProvider;
import com.ferg.awfulapp.reply.Reply;
import com.ferg.awfulapp.thread.AwfulMessage;
import com.ferg.awfulapp.util.AwfulError;
import org.jsoup.nodes.Document;

import java.sql.Timestamp;

/**
 * Created by matt on 8/8/13.
 */
public class QuoteRequest extends AwfulRequest<ContentValues> {
    private int threadId, postId;
    public QuoteRequest(Context context, int threadId, int postId) {
        super(context, Constants.FUNCTION_POST_REPLY);
        this.threadId = threadId;
        this.postId = postId;
    }

    @Override
    protected String generateUrl(Uri.Builder urlBuilder) {
        urlBuilder.appendQueryParameter(Constants.PARAM_ACTION, "newreply");
        urlBuilder.appendQueryParameter(Constants.PARAM_POST_ID, Integer.toString(postId));
        return urlBuilder.build().toString();
    }

    @Override
    protected ContentValues handleResponse(Document doc) throws AwfulError {
        ContentValues newReply = Reply.processQuote(doc, threadId, postId);
        newReply.put(AwfulProvider.UPDATED_TIMESTAMP, new Timestamp(System.currentTimeMillis()).toString());
        return newReply;
    }

    @Override
    protected boolean handleError(AwfulError error, Document doc) {
        return error.isCritical();
    }
}

File: Awful.apk/src/main/java/com/ferg/awfulapp/task/QuoteRequest.kt
package com.ferg.awfulapp.task

import android.content.ContentValues
import android.content.Context
import com.ferg.awfulapp.constants.Constants.*
import com.ferg.awfulapp.provider.DatabaseHelper
import com.ferg.awfulapp.reply.Reply
import com.ferg.awfulapp.util.AwfulError
import org.jsoup.nodes.Document
import java.sql.Timestamp

/**
 * Request the data you get by quoting a post on the main site.
 *
 * This provides you with the text and BBCode added to the reply composer,
 * the form key and cookie (for authentication?) as well as any selected
 * options (see [Reply.processQuote]) and a current timestamp.
 */
class QuoteRequest(context: Context, private val threadId: Int, private val postId: Int)
    : AwfulRequest<ContentValues>(context, FUNCTION_POST_REPLY) {

    init {
        with(parameters) {
            add(PARAM_ACTION, "newreply")
            add(PARAM_POST_ID, postId.toString())
        }
    }

    @Throws(AwfulError::class)
    override fun handleResponse(doc: Document): ContentValues {
        return Reply.processQuote(doc, threadId, postId).apply {
            put(DatabaseHelper.UPDATED_TIMESTAMP, Timestamp(System.currentTimeMillis()).toString())
        }
    }

}
